### PR TITLE
feat: add execroot & runfiles symlink guards to js_binary

### PR DESCRIPTION
**Pull request by gregmagolan**
Fixes https://github.com/aspect-build/rules_js/issues/114

The fs patches here have been modified from their original implementation in rules_nodejs https://github.com/bazelbuild/rules_nodejs/blob/173a0252f273de48a6a237518e9cb0d9854cb8e6/packages/node-patches/src/fs.ts.

The have been improved to allow following symlinks within the sandbox, which is required for finding deps in a symlinks node_modules structure. Since the sandbox (on macos) is a tree of symlinks that all point back to their counterparts outside of the sandbox, the old implementation prevented following symlinks entirely (akin to passing `--preserve-symlinks) since every follow was an escape from the guarded sandbox.

In this implementation, if the readlink escape is a symlink to a symlink then we readlink the escaped symlink. For lstat we do the same. For realpath, if we the relative path from the link to the realpath can be remapped into the escaped root then it is.

In rules_js, the guards are setup to protect the execroot (which protects the sandbox when there is one and protects from escaping into the source tree when there isn't) and to protect the runfiles tree for the js_binary in the build, test or run target since that is a symlink tree as well. This is simpler than what was done in rules_nodejs which was to protect all of the dynamically linked node_modules trees.

### CODE REVIEW

File: js/private/node-patches/src/fs.ts [hunk 1]
```diff
@@ -0,0 +1,671 @@
...
+            }
+        }
+        return str
+    }
+
+    // =========================================================================
+    // fs.readlink
+    // =========================================================================
+
+    // tslint:disable-next-line:no-any
```
Comment (alexeagle) [inline js/private/node-patches/src/fs.ts:327]: may as well toss the tslint lines that are no longer doing anything

**Review by alexeagle — COMMENTED** [review]
let's add the tests too, otherwise it's too tempting to leave it for later...

File: js/private/node-patches/src/fs.ts (hunk 1 above)
Comment (gregmagolan) [reply js/private/node-patches/src/fs.ts:325]: roger roger

**Comment by gregmagolan** [comment]
> let's add the tests too, otherwise it's too tempting to leave it for later...

Done. Brought unit tests in from rules_nodejs for the patches

File: e2e/npm_link_package/src/main.js [hunk 2]
```diff
@@ -15,7 +15,7 @@ console.log(c.id())
 const fp = require('@e2e/lib')
 console.log('--@e2e/lib--')
 console.log(fp.id())
-const rulesFooA = require('../external/rules_foo/foo/node_modules/@aspect-test/a')
+const rulesFooA = require('../../rules_foo/foo/node_modules/@aspect-test/a')
```
Comment (alexeagle) [inline e2e/npm_link_package/src/main.js:18]: Interesting change needed

**Review by alexeagle — APPROVED** [review]

File: e2e/npm_link_package/src/main.js (hunk 2 above)
Comment (gregmagolan) [reply e2e/npm_link_package/src/main.js:18]: yea, the path here is runfiles relative for this test. no legacy external runfiles :)

**Pull request merged**